### Assignment1: Weather model

In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm
import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
weather_df = pd.read_sql_query('select * from weatherinszeged',con=engine)

engine.dispose()

In [3]:
Y = weather_df['apparenttemperature'] - weather_df['temperature']
weather_df['humidity_winspeed_interaction'] = weather_df['humidity'] * weather_df['windspeed'] 
X = weather_df[['humidity','windspeed', 'humidity_winspeed_interaction']]
X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Mon, 18 Nov 2019   Prob (F-statistic):               0.00
Time:                        07:07:52   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=================================================================================================
                                    coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------
const                             0.0839      0.033      2.511      0.012       0.018       0.149
humidity                          0.1775      0.043      4.133      0.000       0.093       0.262
windspeed                         0.0905      0.002     36.797      0.000       0.086       0.095
humidity_winspeed_interaction    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

This model's F statistic is 16660, and the associated p-value is very close to zero. This means that our features add some information to the reduced model and our model is useful in explaining charges.
The R-squared value of our Weather model is 0.341. This means that our model explains 34.1% of the variance in the output, leaving 65.9% unexplained. This means that there's still plenty room for improvement. 

### Assignment1: House price model

In [4]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import statsmodels.api as sm
import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
hprice = pd.read_sql_query('select * from houseprices',con=engine)

engine.dispose()

In [5]:
hprice = pd.concat([hprice,pd.get_dummies(hprice.street, drop_first=True)], axis=1)

dummy_columns = list(pd.get_dummies(hprice.street, drop_first=True).columns)

hprice[['overallqual', 'overallcond', 'grlivarea'] + dummy_columns].head()

overallqual  overallcond  grlivarea  Pave
0            7            5       1710     1
1            6            8       1262     1
2            7            5       1786     1
3            7            5       1717     1
4            8            5       2198     1

In [6]:
X = hprice[['overallqual', 'overallcond', 'grlivarea'] + dummy_columns]
Y = hprice['saleprice']

X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.714
Model:                            OLS   Adj. R-squared:                  0.713
Method:                 Least Squares   F-statistic:                     909.3
Date:                Mon, 18 Nov 2019   Prob (F-statistic):               0.00
Time:                        07:50:17   Log-Likelihood:                -17630.
No. Observations:                1460   AIC:                         3.527e+04
Df Residuals:                    1455   BIC:                         3.530e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.496e+04   1.85e+04     -5.119      0.000   -1.31e+05   -5.86e+04
overallqual  3.289e+04   1002.200     32.822      0.000    3.09e+04    3.49e+04
overallcond   332.0004   1006.368      0.330      0.742   -1642.088    2306.088
grlivarea      55.9093      2.633     21.237      0.000      50.745      61.073
Pave        -1.138e+04   1.74e+04     -0.652      0.514   -4.56e+04    2.28e+04
==============================================================================
Omnibus:                      341.836   Durbin-Watson:                   1.981
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8745.441
Skew:                           0.467   Prob(JB):                         0.00
Kurtosis:                      14.954   Cond. No.                     3.58e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.58e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

This model's F statistic is 909.3, and the associated p-value is very close to zero. This means that our features add some information to the reduced model and our model is useful in explaining house prices. The R-squared value of our Weather model is 0.714 and the Adjusted R-square value is 0.713. This means that our model explains 71.3% of the variance in the output, leaving 28.7% unexplained, there's still room for improvement.